_drafts/best-and-worst-gcc-clang-compiler-flags.md

```diff
@@ -0,0 +1,892 @@
+embedded development with practical examples of why the flags are good or bad"
+image: /img/compiler-flags/example-warnings.png
+author: chris
+---
+
+Compilers have hundreds of flags and configuration settings which can be toggled to control
+performance optimizations, code size, error checks and diagnostic information emitted. Often
+these settings wind up being copy and pasted from project to project, makefile to makefile, but
+it's good to periodically audit the current options selected for a project.
+
+<!-- excerpt start -->
+
+In this article we will explore some of the best and worst compiler flags for GCC (and Clang). Our
+focus will be on flags used for embedded projects but the reasoning applies to other development environments as well. We will explore the impact
+each flag has by walking through practical C code examples.
+
+<!-- excerpt end -->
+
+_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest posts
+straight to your mailbox_
+
+{:.no_toc}
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+{:.no_toc}
+
+### Compiling Example Code
+
+If you would like to follow along compiling the examples presented in this article, all the
+code can be found on Github[^9]. To compile the examples, you will need the 8.3.1 GNU Arm Embedded[^10]
+toolchain and `make` installed on your computer.
+
+## The Best Aggregate Warning Options
+
+Compiler warnings are the first line of defense against catching various errors that could arise at
+runtime and enforcing conventions that lead to a more readable codebase. It is _a lot_ easier to fix
+issues at compilation time than when unexpected behavior is encountered on your platform.
+
+GCC and Clang have several warning flags which will enable a collection of useful checks which we
+will explore in more detail below.
+
+> NOTE: When enabling warning flags for a project that hasn't used them previously, there will likely be _a
+> ton_ of warnings. I'd recommend taking an incremental approach when transitioning the project by only compiling parts of your
+> system (i.e new files) with a more stringent set of warning flags or enabling and fixing sets of
+> warnings at a time. Trying to fix thousands of warnings all at the same time is a recipe for
+> introducing regressions or upsetting fellow developers trying to review your change.
+
+### -Wall & -Wextra
+
+The compiler flag `-Wall` enables a base set of warnings generally agreed upon as being useful and easy to fix. `-Wextra` enables an additional set of flags not covered by `-Wall`
+
+The exact list of flags varies between compiler versions but can easily be found by consulting the GCC documentation for the compiler version you are using[^1].
+
+Here's a c snippet that will happily compile but has a couple unfortunate bugs.
+
+```c
+static void prv_various_memset_bugs(void) {
+  #define NUM_ITEMS 10
+  uint32_t buf[NUM_ITEMS];
+  memset(buf, NUM_ITEMS, 0x0);
+  memset(buf, 0x0, NUM_ITEMS);
+  }
```

Comment:
I believe most of your braces in your code examples should be "tabbed left" once.

Reply:
gah, auto-formatter keeps doing that, looks like I only missed correcting 2 :)